File: kakaoInternship2020/bin/kakaoInternship2020/_03_보석쇼핑/solution.js
function solution(gems){
  var count = new Set(gems).size;
  // 보석 종류가 몇개인지 
  var gemMap = new Map()
  // 보석 종류 => 보석 자리를 저장하기 위한 맵
  var gemLength = []
  // 보석을 모두 포함하는 구간을 저장할 배열 
  gems.forEach((gem, i)=> {
    gemMap.delete(gem)
    gemMap.set(gem, i)
    if(gemMap.size === count){
        
        let s=gemMap.values().next().value;
      gemLength.push([s+1, i+1])
    }
  })
  // console.log(gems.length);
    // console.log(gemLength);
  gemLength.sort((a,b)=> {
    if(a[1]-a[0] === b[1]-b[0]){
      return a[1]-b[1]
    }
    return (a[1]-a[0])-(b[1]-b[0])
  })
  
  return gemLength[0]
}

File: kakaoInternship2020/src/kakaoInternship2020/_03_보석쇼핑/Solution.java
package kakaoInternship2020._03_보석쇼핑;

import java.util.*;
import java.util.stream.*;

class Solution {
    public int[] solution(String[] gems) {
        int[] answer = {0,0};
        
        Set<String> gem_set = Arrays.stream(gems)
                                .map(i->(String)i)
                                .collect(Collectors.toSet());
        Map<String,Integer> gem_map = new HashMap<>();
        int left=0,right=0;
        int min_len=Integer.MAX_VALUE;
        while(left<=right && right<=gems.length){
            
            if(gem_set.size()>gem_map.keySet().size()){
                if(right<=gems.length-1)
                gem_map.put(gems[right],gem_map.getOrDefault(gems[right],0)+1);
                right++;
            }
            
            else if(gem_set.size()==gem_map.keySet().size()){
                
                
                gem_map.put(gems[left],gem_map.getOrDefault(gems[left],0)-1);
                
                if(gem_map.getOrDefault(gems[left],0)<=0)
                    gem_map.remove(gems[left]);
                left++;
                if(min_len>right-left){
                    min_len = right-left;
                    answer[0]=left;
                    answer[1]=right;
                }
            }
            
        }
        
        return answer;
    }
}
